R/tissue_specificity_score.R

```diff
@@ -0,0 +1,162 @@
+                  is.matrix(x)
+                  is.numeric(thresh)
+                  length(thresh) == 1L
+                  is.logical(na.rm)
+                  length(na.rm) == 1L
+              })
+              
+              # aggregate observations by group
+              if (!is.null(group)) {
+                  stopifnot(exprs = {
+                      is.character(group) || is.factor(group) || is.numeric(group)
+                      length(group) == ncol(x)
+                  })
+                  message("aggregating observations by group")
+                  if (is.character(group) || is.numeric(group)) {
+                      group <- factor(as.character(group),
+                                      levels = unique(as.character(group)))
+                  }
+                  ml <- max(table(group))
+                  S <- do.call(cbind, lapply(split(seq.int(ncol(x)), group),
+                                             function(y) {
+                                                 c(y, rep(NA, ml - length(y)))
+                                             }))
+                  x <- rowAvgsPerColSet(X = x, S = S)
+              }
+              
+              # calculate expression specificity score
+              message("calculating '", method, "' for ", nrow(x), " genes and ",
+                      ncol(x), " groups")
+              scores <- switch(method,
+                  tau = rowSums2(1 - (x / rowMaxs(x, na.rm = na.rm)), na.rm = na.rm) / (ncol(x) - 1),
+                  TSI = rowMaxs(x, na.rm = na.rm) / rowSums2(x, na.rm = na.rm),
+                  counts = rowSums2(x > thresh, na.rm = na.rm)
+              )
+              return(scores)
+          })
+
+
+#' @rdname specificityScore-methods
+#' @aliases specificityScore,SummarizedExperiment-method
+setMethod("specificityScore",
+          c("SummarizedExperiment"),
+          function(x, method = c("tau", "TSI", "counts"), group = NULL,
+                   thresh = 0, expr_values = "logcounts", na.rm = FALSE) {
+              ## check expr_values
+              if (is.numeric(expr_values)) {
+                  stopifnot(exprs = {
+                      length(expr_values) == 1L
+                      expr_values <= length(assays(x))
+                  })
+              } else if (is.character(expr_values)) {
+                  stopifnot(exprs = {
+                      length(expr_values) == 1L
+                      expr_values %in% assayNames(x)
+                  })
+              } else {
+                  stop("'expr_values' is not a valid value for use in assay()")
+              }
+              ## extract expression matrix
+              xx <- as.matrix(assay(x, expr_values))
+              ## call matrix method
+              specificityScore(xx, method, group, thresh, expr_values, na.rm)
+          })
+
+
+#' @rdname specificityScore-methods
+#' @aliases specificityScore,SingleCellExperiment-method
+setMethod("specificityScore",
```

Comment:
Not sure this is needed - the SummarizedExperiment method should capture all SE-derived objects, right? And if it's included, I think we could just use `callNextMethod()` directly instead of repeating the code here.

Reply:
You are correct - I removed the `SingleCellExperiment` method.